### PR TITLE
Feature/firebase

### PR DESCRIPTION
**Pull request by Ph0tonic**
New features:
- Firebase connection
- Architecture for firebase and data
- Live heatmap
- Live markers

### CODE REVIEW

File: app/src/main/java/ch/epfl/sdp/database/dao/FirebaseGroupDao.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+//                }
+//
+//                override fun onCancelled(error: DatabaseError) {
+//                    Timber.w("Failed to read group from firebase.")
+//                }
+//            })
+//        }
+//        return watchedGroupsById[groupId]!!
+//    }
+//}
```
Comment (Nchappui) [inline app/src/main/java/ch/epfl/sdp/database/dao/FirebaseGroupDao.kt:49]: This file is entirely empty, is it necessary?

File: app/src/main/java/ch/epfl/sdp/database/dao/FirebaseHeatmapDao.kt [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import com.google.firebase.database.ktx.database
+import com.google.firebase.ktx.Firebase
+import timber.log.Timber
+
+class FirebaseHeatmapDao : HeatmapDao {
+
+    private var database: FirebaseDatabase = Firebase.database
+
+    // group_id / user_id / heatmap_data
+    val groupHeatmaps: MutableMap<String, MutableLiveData<MutableMap<String, MutableLiveData<HeatmapData>>>> = mutableMapOf()
```
Comment (Sporarum) [inline app/src/main/java/ch/epfl/sdp/database/dao/FirebaseHeatmapDao.kt:19]: Couldn't this be replaced by a MutableMap<String, MutableLiveData<MutableMap<String,HeatmapData>>> ?
Where the updates would all be handled by the upper layer

File: app/src/main/java/ch/epfl/sdp/database/data/SearchGroupData.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+        @get:PropertyName("base_location")
+        @set:PropertyName("base_location")
+        var baseLocation: LatLng? = null,
+
+        var name: String = "",
+
+        @get:PropertyName("search_location")
+        @set:PropertyName("search_location")
+        var searchLocation:LatLng? = null){
```
Comment (Nchappui) [inline app/src/main/java/ch/epfl/sdp/database/data/SearchGroupData.kt:17]: These four variables aren't used anywhere, are they for a future use?

File: app/src/main/java/ch/epfl/sdp/database/data/UserData.kt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package ch.epfl.sdp.database.data
+
+import com.google.firebase.database.PropertyName
+
+data class UserData(
+        @get:PropertyName("google_id")
+        @set:PropertyName("google_id")
+        var googleId: String = ""
```
Comment (Nchappui) [inline app/src/main/java/ch/epfl/sdp/database/data/UserData.kt:8]: Same here

File: app/src/main/java/ch/epfl/sdp/database/dao/FirebaseHeatmapDao.kt (hunk 2 above)
Comment (Ph0tonic) [reply app/src/main/java/ch/epfl/sdp/database/dao/FirebaseHeatmapDao.kt:19]: No, we have a heatmap painter for each heatmap and this allows us to make it observe directly the right data. Therefore, we don't need to detect which data changes and forwards the event to the right heatmap painter.

File: app/src/main/java/ch/epfl/sdp/database/dao/FirebaseGroupDao.kt (hunk 1 above)
Comment (Ph0tonic) [reply app/src/main/java/ch/epfl/sdp/database/dao/FirebaseGroupDao.kt:49]: It going to be used for the work of this week.

File: app/src/main/java/ch/epfl/sdp/database/data/SearchGroupData.kt (hunk 3 above)
Comment (Ph0tonic) [reply app/src/main/java/ch/epfl/sdp/database/data/SearchGroupData.kt:17]: Yes

File: app/src/main/java/ch/epfl/sdp/database/data/UserData.kt (hunk 4 above)
Comment (Ph0tonic) [reply app/src/main/java/ch/epfl/sdp/database/data/UserData.kt:8]: Idem

**Pull request merged**